mlflow2seldon/api.py

```diff
@@ -0,0 +1,291 @@
+                            deploy_image
+                        ).as_docker_url()
+                    else:
+                        deploy_image_ref = default_image_ref
+                    for model_version in model.latest_versions:
+                        if model_version.current_stage not in ("Staging", "Production"):
+                            # we deploy only Staging and Production models
+                            continue
+                        # Assumption: artifact store in MLflow is platform storage
+                        # mlflow-config related path, e.g.
+                        # ('/', 'usr', 'local', 'share', 'mlruns', '0',
+                        # 'ae72265a0a17473f993f78ab239c2f2f', 'artifacts', 'model')
+                        source_path_parts = Path(model_version.source).parts
+                        # leave pure mlflow subpath, e.g.:
+                        # (0, ae72265a0a17473f993f78ab239c2f2f, artifacts, model)
+                        mlruns_idx = source_path_parts.index(model_version.run_id) - 1
+                        mlflow_source_parts = source_path_parts[mlruns_idx:]
+                        storage_art_uri = URL(
+                            f"{mlflow_storage_root}/{'/'.join(mlflow_source_parts)}"
+                        )
+                        registered_model = _DeployedModel(
+                            image=deploy_image_ref,
+                            model_storage_uri=storage_art_uri,
+                            model_name=model.name,
+                            model_stage=model_version.current_stage,
+                            model_version=model_version.version,
+                            source_run_id=model_version.run_id,
+                            deployment_namespace=seldon_deployment_ns,
+                        )
+                        deployed_model = seldon_models.get(registered_model.name)
+                        if not deployed_model or not deployed_model.is_same_version(
+                            registered_model
+                        ):
+                            registered_model.need_redeploy = True
+                        else:
+                            registered_model.need_redeploy = False
+                        mlflow_models[registered_model.name] = registered_model
+
+                # deploy models in Seldon
+                for model_name in mlflow_models.keys():
+                    if mlflow_models[model_name].need_redeploy:
+                        await _deploy_model(
+                            mlflow_models[model_name],
+                            neuro_client,
+                            registry_secret_name,
+                        )
+                        mlflow_models[model_name].need_redeploy = False
+
+                # removing outdated models
+                for model_name in set(seldon_models.keys()) - set(mlflow_models.keys()):
+                    _delete_seldon_deployment(seldon_models[model_name])
```

Comment:
hmm let's better drop the whole try-finally: It's not logically clean if the Operator (the one that syncs the state MLflow -> Seldon) deletes models registered in MLflow in Seldon once it finishes its work. It's better that the Operator just stops synching once it terminates.

Reply:
sounds reasonable, changed in 743a444